src/SolverAlsWrmf.cpp

```diff
@@ -151,27 +292,123 @@ double als_implicit_float(const Rcpp::S4 &m_csc_r,
+                           arma::Col<double>& m_values_orig,
+                           arma::mat& X,
+                           arma::mat& Y,
+                           double lambda,
+                           unsigned n_threads,
+                           unsigned solver,
+                           unsigned cg_steps = 3,
+                           bool calc_user_bias = false,
+                           bool calc_item_bias = false,
+                           bool non_negative = false) {
+  const dMappedCSC Conf = extract_mapped_csc(m_csc_r);
+  return (double)als_explicit_cpp<double>(Conf, m_values_orig,
+                                          X, Y, lambda, n_threads,
+                                          solver, cg_steps,
+                                          calc_user_bias, calc_item_bias,
+                                          non_negative);
+}
 
 // [[Rcpp::export]]
-double als_loss_explicit(const Rcpp::S4 &m_csc_r, arma::mat& X, arma::mat& Y, double lambda, unsigned n_threads) {
-  dMappedCSC mat = extract_mapped_csc(m_csc_r);
-  size_t nc = mat.n_cols;
-  double loss = 0;
-  #ifdef _OPENMP
-  #pragma omp parallel for num_threads(n_threads) schedule(dynamic, GRAIN_SIZE) reduction(+:loss)
-  #endif
-  for(size_t i = 0; i < nc; i++) {
-    arma::colvec y_i = Y.col(i);
-    auto p1 = mat.col_ptrs[i];
-    auto p2 = mat.col_ptrs[i + 1];
-    double diff = 0;
-    for(int pp = p1; pp < p2; pp++) {
-      auto j = mat.row_indices[pp];
-      diff = mat.values[pp] - arma::dot(y_i, X.col(j));
-      loss += diff * diff;
+double als_explicit_float(const Rcpp::S4 &m_csc_r,
+                          arma::Col<double>& m_values_orig,
+                          Rcpp::S4 &XR,
+                          Rcpp::S4 & YR,
+                          double lambda,
+                          unsigned n_threads,
+                          unsigned solver,
+                          unsigned cg_steps = 3,
+                          bool calc_user_bias = false,
+                          bool calc_item_bias = false,
+                          bool non_negative = false) {
+  //#ifdef SINGLE_PRECISION_LAPACK_AVAILABLE
+  const dMappedCSC Conf = extract_mapped_csc(m_csc_r);
+  Rcpp::IntegerMatrix XRM = XR.slot("Data");
+  Rcpp::IntegerMatrix YRM = YR.slot("Data");
+  float * x_ptr = reinterpret_cast<float *>(&XRM[0]);
+  float * y_ptr = reinterpret_cast<float *>(&YRM[0]);
+  arma::fmat X = arma::fmat(x_ptr, XRM.nrow(), XRM.ncol(), false, true);
+  arma::fmat Y = arma::fmat(y_ptr, YRM.nrow(), YRM.ncol(), false, true);
+  return (double)als_explicit_cpp<float>(Conf, m_values_orig,
+                                         X, Y, lambda, n_threads,
+                                         solver, cg_steps,
+                                         calc_user_bias, calc_item_bias,
+                                         non_negative);
+}
+
+template <class T>
+void initialize_biases(const dMappedCSC& ConfCSC,
```

Comment:
@david-cortes  I believe it will be easier/more efficient to pass `global_mean` here and to ALS solver rather then subtracting it from `c_ui` and `c_iu`.